sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java

```diff
@@ -24,308 +25,375 @@
+                            }));
     }
 
-    @Override
-    public boolean equals(Object obj) {
-      if (!(obj instanceof TestMatcher)) {
-        return false;
+    /**
+     * Tests for {@link TestPipeline}s with a non {@link CrashingRunner}.
+     */
+    @RunWith(JUnit4.class)
+    public static class WithRealPipelineRunner {
+
+      private final transient ExpectedException exception = ExpectedException.none();
+
+      private final transient TestPipeline pipeline = TestPipeline.create();
+
+      @Rule
+      public final transient RuleChain chain = RuleChain.outerRule(exception).around(pipeline);
+
+      @Category(RunnableOnService.class)
+      @Test
+      public void testNormalFlow() throws Exception {
+        addTransform(pCollection(pipeline));
+        pipeline.run();
       }
-      TestMatcher other = (TestMatcher) obj;
-      return other.uuid.equals(uuid);
-    }
 
-    @Override
-    public int hashCode() {
-      return uuid.hashCode();
-    }
-  }
+      @Category(RunnableOnService.class)
+      @Test
+      public void testMissingRun() throws Exception {
 
-  private static class DummyRunner extends PipelineRunner<PipelineResult> {
+        exception.expect(TestPipeline.PipelineRunMissingException.class);
 
-    @SuppressWarnings("unused") // used by reflection
-    public static DummyRunner fromOptions(final PipelineOptions opts) {
-      return new DummyRunner();
-    }
+        addTransform(pCollection(pipeline));
+      }
 
-    @Override
-    public PipelineResult run(final Pipeline pipeline) {
-      return new PipelineResult() {
+      @Category(RunnableOnService.class)
+      @Test
+      public void testMissingRunWithDisabledEnforcement() throws Exception {
+        pipeline.enableAbandonedNodeEnforcement(false);
+        addTransform(pCollection(pipeline));
 
-        @Override
-        public State getState() {
-          return null;
-        }
+        // disable abandoned node detection
+      }
 
-        @Override
-        public State cancel() throws IOException {
-          return null;
-        }
+      @Category(RunnableOnService.class)
+      @Test
+      public void testMissingRunAutoAdd() throws Exception {
+        pipeline.enableAutoRunIfMissing(true);
+        addTransform(pCollection(pipeline));
 
-        @Override
-        public State waitUntilFinish(final Duration duration) {
-          return null;
-        }
+        // have the pipeline.run() auto-added
+      }
 
-        @Override
-        public State waitUntilFinish() {
-          return null;
-        }
+      @Category(RunnableOnService.class)
+      @Test
+      public void testDanglingPTransformRunnableOnService() throws Exception {
 
-        @Override
-        public <T> AggregatorValues<T> getAggregatorValues(final Aggregator<?, T> aggregator)
-            throws AggregatorRetrievalException {
-          return null;
-        }
+        exception.expect(TestPipeline.AbandonedNodeException.class);
```

Comment:
I would move the expectations to occur immediately before the throwing statement.